components/tikv_util/src/mpsc/batch.rs

```diff
@@ -422,33 +446,46 @@ mod tests {
 
     #[test]
     fn test_switch_between_sender_and_receiver() {
-        struct Notifier<F>(Arc<Mutex<Option<Spawn<F>>>>);
-        impl<F> Clone for Notifier<F> {
-            fn clone(&self) -> Notifier<F> {
-                Notifier(Arc::clone(&self.0))
-            }
-        }
-        impl<F> Notify for Notifier<F>
-        where
-            F: Future<Item = (), Error = ()> + Send + 'static,
-        {
-            fn notify(&self, id: usize) {
-                let n = Arc::new(self.clone());
-                let mut s = self.0.lock().unwrap();
-                match s.as_mut().map(|spawn| spawn.poll_future_notify(&n, id)) {
-                    Some(Ok(Async::NotReady)) | None => {}
-                    _ => *s = None,
+        let (tx, mut rx) = unbounded::<i32>(4);
+        let future = async move { rx.next().await };
+        let mut task = Task {
+            future: Arc::new(Mutex::new(Some(future.boxed()))),
+        };
+        // Receiver has not received any messages, so the future is not be finished
+        // in this tick.
+        task.tick();
+        assert!(task.future.lock().unwrap().is_some());
+        // After sender is dropped, waker will send the task to executor again and
+        // the future will be finished in this tick.
+        drop(tx);
+        task.tick();
+        assert!(task.future.lock().unwrap().is_none());
+    }
+
+    #[derive(Clone)]
+    struct Task {
+        future: Arc<Mutex<Option<BoxFuture<'static, Option<i32>>>>>,
+    }
+
+    impl Task {
+        fn tick(&mut self) {
+            let task = Arc::new(self.clone());
+            let mut future_slot = self.future.lock().unwrap();
+            if let Some(mut future) = future_slot.take() {
+                let waker = task::waker_ref(&task);
+                let cx = &mut Context::from_waker(&*waker);
+                match future.as_mut().poll(cx) {
+                    Poll::Pending => {
+                        *future_slot = Some(future);
+                    }
+                    Poll::Ready(None) => {}
+                    _ => unimplemented!(),
                 }
             }
         }
+    }
 
-        let (tx, rx) = unbounded::<u64>(4);
-        let f = rx.for_each(|_| Ok(())).map_err(|_| ());
-        let spawn = Arc::new(Mutex::new(Some(executor::spawn(f))));
-        Notifier(Arc::clone(&spawn)).notify(0);
-
-        // Switch to `receiver` in one thread in where `sender` is dropped.
-        drop(tx);
-        assert!(spawn.lock().unwrap().is_none());
+    impl ArcWake for Task {
+        fn wake_by_ref(_: &Arc<Self>) {}
```

Comment:
This should call `tick()`. Otherwise you can't prove the task is scheduled when it's ready.